go/inst/oracle_gtid_set_entry.go

```diff
@@ -24,51 +24,233 @@ import (
+//		- "230ea8ea-81e3-11e4-972a-e25ec4bd140a:1-5139:5780-6317:tag1:1-2474:3201-4157" (tagged intervals and may have gaps)
 type OracleGtidSetEntry struct {
-	UUID   string
-	Ranges string
+	UUID      string
+	DefaultIv string        // default (untagged) interval
+	TaggedIv  []TagInterval // tagged intervals
 }
 
-// NewOracleGtidSetEntry parses a single entry text
-func NewOracleGtidSetEntry(gtidRangeString string) (*OracleGtidSetEntry, error) {
-	gtidRangeString = strings.TrimSpace(gtidRangeString)
-	tokens := strings.SplitN(gtidRangeString, ":", 2)
-	if len(tokens) != 2 {
+func ParseOracleGtidSetEntry(gtidRangeString string) (*OracleGtidSetEntry, error) {
+
+	// Split the string into two parts: UUID part and the non-UUID part
+	gtid_str := strings.SplitN(gtidRangeString, ":", 2)
+
+	// Sanity check
+	if len(gtid_str) != 2 {
 		return nil, fmt.Errorf("Cannot parse OracleGtidSetEntry from %s", gtidRangeString)
 	}
-	if tokens[0] == "" {
-		return nil, fmt.Errorf("Unexpected UUID: %s", tokens[0])
+
+	if gtid_str[0] == "" {
+		return nil, fmt.Errorf("Unexpected UUID: %s", gtid_str[0])
+	}
+
+	if gtid_str[1] == "" {
+		return nil, fmt.Errorf("Unexpected GTID range: %s", gtid_str[1])
+	}
+
+	// UUID is the first part
+	uuid := gtid_str[0]
+
+	// Split the non-UUID parts into multiple blocks
+	s := strings.SplitN(gtid_str[1], ":", -1)
+
+	// Initialize the tag and interval
+	var default_iv string     // Default interval
+	var tag_ivs []TagInterval // Full tagged interval
+	var ti TagInterval        // Current tag interval
+
+	for i := range s {
+
+		// If it is a GTID tag
+		if tagRegex.MatchString(s[i]) {
+
+			if (ti.Tag != "") && (len(ti.Interval) == 0) {
+				// If the tag is already set and we got another tag
+				return nil, fmt.Errorf("Invalid format")
+			} else if (ti.Tag == "") && (len(ti.Interval) != 0) {
+				// If the tag is not set and we already have the interval set
+				return nil, fmt.Errorf("Invalid format")
+			} else {
+				// Now process the new tag
+				ti.Tag = s[i]
+				// Reset iv for the current tag
+				ti.Interval = nil
+				// Append the new tag to tag_ivs
+				tag_ivs = append(tag_ivs, ti)
+			}
+		} else {
+			// If it is an GTID interval
+			if singleValueInterval.MatchString(s[i]) || multiValueInterval.MatchString(s[i]) {
+				// If it is an empty tag, add it to default interval
+				if len(ti.Tag) == 0 {
+					default_iv += ":" + s[i]
+				} else {
+					// If tag is already set, add it to the tag interval
+					ti.Interval = append(ti.Interval, s[i])
+					tag_ivs[len(tag_ivs)-1].Interval = append(tag_ivs[len(tag_ivs)-1].Interval, s[i])
+				}
+			} else {
+				// Regex failed, invalid format
+				return nil, fmt.Errorf("Invalid format")
+			}
+		}
+	}
+
+	// If the interval of the last tag is empty, then it is an invalid format
+	// eg: "UUID:1-5139::tag1:"
+	if (ti.Tag != "") && (len(tag_ivs[len(tag_ivs)-1].Interval) == 0) {
+		return nil, fmt.Errorf("Invalid format")
```

Comment:
```suggestion
		// Initialize the tag and interval
	var default_iv string      // Default interval
	var tag_ivs []TagInterval  // Full tagged interval
	var tip *TagInterval = nil // Current tag interval
	// Tagged intervals always follow untagged ones
	// so once tip != nil it will never be nil again

	for i := range s {

		// If it is a GTID tag
		if tagRegex.MatchString(s[i]) {

			if tip != nil && (tip.Tag != "") && (len(tip.Interval) == 0) {
				// If the tag is already set and we got another tag
				return nil, fmt.Errorf("Invalid format")
			} else if tip != nil && (tip.Tag == "") && (len(tip.Interval) != 0) {
				// If the tag is not set and we already have the interval set
				// KH: is it possible at all? To set Interval, we need non empty
				// Tag (below)
				return nil, fmt.Errorf("Invalid format")
			} else {
				// Now process the new tag
				ti := TagInterval{
					Tag:      s[i],
				}

				// Append the new tag to tag_ivs
				tag_ivs = append(tag_ivs, ti)
				tip = &tag_ivs[len(tag_ivs)-1]
			}
		} else {
			// If it is an GTID interval
			if singleValueInterval.MatchString(s[i]) || multiValueInterval.MatchString(s[i]) {
				// If it is an empty tag, add it to default interval
				if tip == nil {
					default_iv += ":" + s[i]
				} else {
					// If tag is already set, add it to the tag interval
					tip.Interval = append(tip.Interval, s[i])
				}
			} else {
				// Regex failed, invalid format
				return nil, fmt.Errorf("Invalid format")
			}
		}
	}

	// If the interval of the last tag is empty, then it is an invalid format
	// eg: "UUID:1-5139::tag1:"
	if tip != nil && (tip.Tag != "") && (len(tip.Interval) == 0) {
		return nil, fmt.Errorf("Invalid format")
	}

	// Don't append ':' for the first interval in the default set
	default_iv, _ = strings.CutPrefix(default_iv, ":")

	entry := OracleGtidSetEntry{UUID: uuid, DefaultIv: default_iv, TaggedIv: tag_ivs}

	return &entry, nil
```

Reply:
Done